### PR TITLE
fix(naming): stop a C# sibling type being named as a member

### PR DESCRIPTION
**Pull request by Svilen-Stefanov**
Last of three, now based on `main` with #543 and #544 merged.

C# has no file scope, so a file may declare several top-level types. The adapter folded the file stem into the module and then collapsed the type whose name matched it, which left every *other* top-level type named as a **member of the type the file is named after**:

```
Models/Animal.cs  declaring Animal, Dog and Cat as siblings

  was   Models.Animal,        Models.Animal.Dog,  Models.Animal.Cat
  now   Models.Animal.Animal, Models.Animal.Dog,  Models.Animal.Cat
```

This was not cosmetic:

- `internal_references.is_self_or_container_edge` tests dot-prefixes, so a real dependency between two types in one file read as containment and was deleted.
- CONTAINS attached a sibling to a class that does not declare it.
- It **destroyed a symbol outright**. `Task.cs` declaring `enum Priority` beside `class Task { Priority Priority }` named both `Models.Task.Priority`, and `SymbolTable._symbols[qname] = info` has no guard.

That last one is demonstrated rather than argued: against the real engine the C# edge-case project goes from **97 references to 98**, and the extra one is `Models.Task.Task.Priority` — the property that used to lose the collision. See CodeBoarding-tests#27.

A file declaring one type — the overwhelming case — is **byte-identical**. `Models/User.cs` still gives `Models.User` and `Models.User.Name`.

## Measured

Share of top-level types that get a new name, and collisions before and after:

| | types | renamed | | collisions before | after |
|---|---|---|---|---|---|
| eShop | 552 | 34 | 6.2% | 0 | 0 |
| abp | 9199 | 337 | 3.8% | 0 | 0 |
| Polly | 941 | 59 | 6.6% | 0 | 0 |
| AutoMapper | 3750 | 53 | 5.4% | 0 | 0 |
| serilog | 262 | 21 | 8.0% | 0 | 0 |

## Why the stem is not simply dropped

Dropping it for every type — the obvious smaller change, and what #540 did first — reads better but loses declarations, because the stem is what separates two same-named types in one directory: eShop 10, abp 160, Polly 105, AutoMapper 7, serilog 1. Keeping it exactly where it is load-bearing costs nothing.

## Why the adapter is handed the symbol tree

What *else* a file declares cannot come from a name built one symbol at a time, so `record_document_symbols` gives the adapter the file's tree once, before its names are built. It is called from the single production registration path, and it both records and discards, so a warm re-analysis of a file that lost a sibling agrees with a cold run. A name stays a pure function of one file, which is what keeps it stable on an incremental run.

## No cache tag bump

v6 already covers this. It landed in #544 and ships in the same release, so the note beside it now names both halves of the C#/Java rename instead of adding a v7 nobody would ever see on its own.

## Verified

`pytest tests/` → 2258 passed (3 pre-existing `test_cli_parser` failures deselected; identical on a clean `main`). `black`, `mypy` clean. Against the real engine, all 9 C# edge-case integration tests pass on the migrated fixtures (CodeBoarding-tests#27), including two edges corrected rather than renamed: `: base(size)` and `new(5)` call `Settings(int size)`, never the parameterless `Settings()` the old fixture pinned.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Improvements**
  - Improved C# code analysis for files containing multiple top-level types, keeping sibling and nested type names distinct.
  - Preserved existing qualified-name behavior for files containing a single top-level type.
  - Re-analysis now updates qualified names when a file’s type structure changes.
  - Enums are correctly accounted for when determining file type structure.

- **Documentation**
  - Updated cache-format documentation to describe qualified-name behavior changes while retaining cache version `v6`.

### CODE REVIEW

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 1]
```diff
@@ -182,6 +188,10 @@ def _ensure_csharp_ls_installed(self, project_root: Path, dotnet_path: str, dotn
                 "to analyze C# projects."
             )
 
+    def record_document_symbols(self, file_path: Path, symbols: list[dict]) -> None:
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py]: Fixed — `record_document_symbols` now sits at the top of the class, above every `_`-prefixed method.

File: static_analyzer/engine/language_adapter.py [hunk 2]
```diff
@@ -356,6 +356,13 @@ def resolves_collection_initializers(self) -> bool:
         once per element, rather than only constructing the type."""
         return False
 
+    def record_document_symbols(self, file_path: Path, symbols: list[dict]) -> None:
+        """Take note of a file's symbol tree before its names are built.
+
+        Why: a name is built one symbol at a time, so an adapter that needs to know what
+        else the file declares has nowhere else to learn it.
+        """
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/language_adapter.py:365]: Fixed — one line plus a short `Why:`.

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 3]
```diff
@@ -182,6 +188,10 @@ def _ensure_csharp_ls_installed(self, project_root: Path, dotnet_path: str, dotn
...
+        if len(self._top_level_types(symbols)) > 1:
+            self._files_with_sibling_types.add(str(file_path))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py]: Agreed. The set is now updated in both directions on every document-symbol response, so a file that loses a sibling goes back to the collapsed name and a warm re-analysis agrees with a cold one. Test added for exactly that sequence.

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 4]
```diff
@@ -398,3 +394,14 @@ def is_reference_worthy(self, symbol_kind: int) -> bool:
...
         return self._get_hierarchical_packages(source_files, project_root)
+
+    def _top_level_types(self, symbols: list[dict]) -> list[str]:
+        """The type declarations a file makes outside any other type."""
+        found: list[str] = []
+        for sym in symbols:
+            kind = sym.get("kind", 0)
+            if kind in (NodeType.FILE, NodeType.NAMESPACE):
+                found.extend(self._top_level_types(sym.get("children", [])))
+            elif self.is_class_like(kind):
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py:412]: I checked this against csharp-ls rather than reasoning about it, and the premise doesn't hold: a top-level delegate is already counted.

Probe — `Sample.cs` declaring `public delegate void Notify(string)` beside `public class Sample`, analysed end to end:

```
Probe
Sample.Notify
Sample.Sample          <- not collapsed, so the file was classified multi-type
Sample.Sample.Run()
```

If the delegate had been excluded the class would have come out as `Sample` and the delegate as `Sample.Notify`, which is the failure you describe. csharp-ls 0.24 reports a delegate with a class-like kind, so `is_class_like` already accepts it. No change made.

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 5]
```diff
@@ -191,42 +201,28 @@ def build_qualified_name(
...
-            if code_parents:
-                return f"{module}.{'.'.join(code_parents)}.{symbol_name}"
-
-        # Deduplicate filename/class (one type per file convention)
-        module_last = module.rsplit(".", 1)[-1] if "." in module else module
-        if symbol_name == module_last:
-            return module
-        return f"{module}.{symbol_name}"
+            if not code_parents and symbol_name == rel.stem:
+                return module
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py]: This is real, and I want to be precise about what it is: it is the pre-existing behaviour of the incremental path for *any* rename, and what this PR adds is a new way to trigger it.

`_restore_inbound_edges_via_definitions` skips an invalidated edge whose destination node no longer exists (`incremental_orchestrator.py:227-228`). That is true today whenever an edit renames a symbol — rename a class on `main` and inbound edges from unchanged files are dropped exactly the same way. This PR makes "add a second top-level type to a file" one more edit that renames the file-named type and its members.

The window is narrow and self-healing: it needs a file that declared exactly one top-level type to gain one, with inbound edges from unchanged files, and the next full analysis restores them. The v7 tag bump also forces a full run on upgrade, so it cannot bite on the first analysis after this lands.

The fix I'd want is general rather than specific to this trigger: that restore already re-resolves definitions from the cached call sites, so it could follow the rename by accepting whatever node the position now resolves to instead of requiring the old name to still exist. That repairs every rename, not just this one, and changes incremental semantics broadly enough to deserve its own PR and its own warm-start test. Filing it.

File: static_analyzer/analysis_cache.py [hunk 6]
```diff
@@ -63,8 +63,10 @@
 # v6: every Java qualified name changes. The Maven/Gradle source root no longer prefixes
 # a name, and the file stem is gone, so a top-level type is no longer doubled and its
 # members can reach it.
+# v7: a C# file declaring more than one top-level type keeps the file stem as a segment, so
+# a sibling is no longer named as a member of the type the file is named after.
 # Older pickles are treated as cache misses and re-run.
-_TAG_VERSION = "v6"
+_TAG_VERSION = "v7"
```
Comment (ivanmilevtues) [inline static_analyzer/analysis_cache.py]: no need to bump version again v6 from prev branch is good enough.

**Review by ivanmilevtues — APPROVED** [review]
blockerish: no need to bump tag_version again v6 is already a bump in the stack.

You can merge once it is v6.

Looks good otherwise.

File: static_analyzer/analysis_cache.py (hunk 6 above)
Comment (Svilen-Stefanov) [reply static_analyzer/analysis_cache.py]: Done — the bump is gone and the branch is rebased on `main` with #543 and #544 in. Rather than leave the v6 note describing only the Java half, it now names both, since the two ship in the same release:

```
# v6: C#/Java qualified names change. For Java the Maven/Gradle source root no longer
# prefixes a name and the file stem is gone ...; for C# a file declaring more than one
# top-level type keeps the stem as a segment ...
```

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 7]
```diff
@@ -118,10 +118,24 @@ def _single_target_framework_env(project_root: Path) -> dict[str, str]:
...
     def include_references_on_declaration_line(self) -> bool:
         return True
 
+    def record_document_symbols(self, file_path: Path, symbols: list[dict]) -> None:
+        # Discarded as well as added: the same adapter can re-analyse a file after a sibling
+        # was deleted, and it must then name it the way a cold run would.
+        if len(self._top_level_types(symbols)) > 1:
+            self._files_with_sibling_types.add(str(file_path))
+        else:
+            self._files_with_sibling_types.discard(str(file_path))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py]: Good catch — real, and I'd introduced it: this is the first mutable state `CSharpAdapter` has ever carried, and line 206 of `static_analyzer/__init__.py` hands the *same* adapter instance to every C# `EngineConfig`.

Fixed by scoping the classification to the engine config rather than synchronising it. `build_qualified_name` already receives `project_root`, so the set is now keyed `(project root, path)` and `record_document_symbols` takes the root too. Two configs holding the same file each keep their own answer, and there is no shared entry left for them to race on — no lock, and no protocol change beyond the hook's own signature.

Test added: the same file recorded as multi-type under one root and single-type under another still names it per root.

File: static_analyzer/engine/adapters/csharp_adapter.py [hunk 8]
```diff
@@ -191,42 +205,28 @@ def build_qualified_name(
...
-                return f"{module}.{'.'.join(code_parents)}.{symbol_name}"
-
-        # Deduplicate filename/class (one type per file convention)
-        module_last = module.rsplit(".", 1)[-1] if "." in module else module
-        if symbol_name == module_last:
-            return module
-        return f"{module}.{symbol_name}"
+            if not code_parents and symbol_name == rel.stem:
+                return module
+        return ".".join((module, *code_parents, symbol_name))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/engine/adapters/csharp_adapter.py:232]: Correct, and it is genuinely new in shape — on `main` the stem-named type collapses, so `A.cs` gives `Dir.A` and `A/A.cs` gives `Dir.A.A`; keeping the stem for a multi-type `A.cs` makes both `Dir.A.A`. I am leaving it, and here is the reasoning rather than a hand-wave.

**Measured: 0 occurrences.** I scanned for the exact shape — a multi-type `X.cs` sitting beside a directory `X/` that contains `X.cs` — across 10,035 C# files:

| eShop | abp | Polly | AutoMapper | serilog |
|---|---|---|---|---|
| 0 | 0 | 0 | 0 | 0 |

**It is not a new *class* of defect.** `csharp_adapter.py:223` drops `src` from the path, so on `main` today two files differing only by a `src/` segment already flatten to one name. Any scheme that folds a filesystem path into dot-separated segments has this property; the dots are not injective over paths.

**Every alternative loses more.** Dropping the stem instead — the fix that removes this collision — silently loses 10 declarations on eShop, 160 on abp, 105 on Polly, 7 on AutoMapper and 1 on serilog. Trading a measured 0 for a measured 283 is the wrong direction.

Where I think you are right is the last clause: *reject the collision rather than returning an incomplete graph*. That belongs at `symbol_table.py:119`, which today does `self._symbols[qualified_name] = info` with no guard at all, and a guard there would catch the `src` case and every other flattening collision, not just this one. It is a change to every language's behaviour on a path this PR does not touch, so I would rather do it deliberately with its own measurement of what it starts rejecting. Filing it.
